src/dataflow/src/sink/tail.rs

```diff
@@ -66,6 +70,29 @@ pub fn tail<G>(
             // though it is slower because it will produce deterministic results since the
             // cursor will always produce rows in the same order.
             results.sort_by_key(|(time, _)| *time);
+
+            if connector.emit_timestamps && !batches.is_empty() {
+                // The user has request timestamp done messages. We've already sorted the
+                // results by timestamp so we can duplicate the last row's timestamp which is
+                // somewhat easier than using a batch.upper() antichain. This timestamp won't
+                // be reused any other batch because they are disjoint and strictly increasing.
+                let (time, _) = results.last().unwrap().clone();
```

Comment:
there might still be multiple timestamps stored in results and if I'm reading this correctly, we would only emit a done message for the largest timestamp in `results` -- I'm not sure if thats the intent but if it is I think we should call that out in the docs more clearly?

Reply:
You are correct. Added a paragraph documenting this.

Reply:
Another option here that communicates more information is to grab `batch.upper()` which will report the upper end of the open interval of timestamps contained in the batch. This doesn't have to be the largest timestamp in the batch; it could be larger.